File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/wheel/GroundFrictionAccelerationCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.wheel;

import com.github.aadvorak.artilleryonline.battle.calculations.WheelCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.WheelGroundState;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;

public class GroundFrictionAccelerationCalculator {

    public static void calculate(WheelCalculations wheelCalculations, VehicleModel vehicleModel,
                                 double groundFrictionCoefficient) {
        if (WheelGroundState.FULL_OVER_GROUND.equals(wheelCalculations.getGroundState())) {
            return;
        }
        var depth = WheelGroundState.FULL_UNDER_GROUND.equals(wheelCalculations.getGroundState())
                ? 2 * vehicleModel.getSpecs().getWheelRadius() : wheelCalculations.getGroundDepth();
        wheelCalculations.getGroundFrictionAcceleration()
                .setX( - wheelCalculations.getVelocity().getX() * depth * groundFrictionCoefficient)
                .setY( - wheelCalculations.getVelocity().getY() * depth * groundFrictionCoefficient);
    }
}

File: playground/src/battle/calculator/wheel/ground-friction-acceleration-calculator.ts
import { type WheelCalculations, WheelGroundState } from '@/data/calculations'
import type { VehicleModel } from '@/data/model'

export const GroundFrictionAccelerationCalculator = {
  calculate(
    wheelCalculations: WheelCalculations,
    vehicleModel: VehicleModel,
    groundFrictionCoefficient: number
  ): void {
    if (WheelGroundState.FULL_OVER_GROUND === wheelCalculations.groundState) {
      return
    }

    const depth =
      WheelGroundState.FULL_UNDER_GROUND === wheelCalculations.groundState
        ? 2 * vehicleModel.specs.wheelRadius
        : wheelCalculations.groundDepth!

    wheelCalculations.groundFrictionAcceleration.x =
      -wheelCalculations.velocity!.x * depth * groundFrictionCoefficient
    wheelCalculations.groundFrictionAcceleration.y =
      -wheelCalculations.velocity!.y * depth * groundFrictionCoefficient
  }
}
